logmeinUrl package and usage

lsl opened the pull request:
#### Changes proposed in this Pull Request

* Add `logmeingUrl` method to append the logmein parameter defined in D61461-code

#### Testing instructions

* ~Apply D61461-code~ (dependency, since deployed)
* Login to http://calypso.localhost:3000/, create a private site with a mapped domain (or log into my test user for `testing23465324891.home.blog/`)
* oauth flag set for testing without 3pc, and logmein flag set for testing this feature, e.g. visit `http://calypso.localhost:3000/?flags=oauth,logmein`
* Click here: ![image](https://user-images.githubusercontent.com/5436027/118505955-84e51280-b74a-11eb-8a6c-da4797a165ee.png)

You should end up the site in a logged in state with cookies against the mapped domain.

- Test both while logged out of the test site directly.
- Test sub.subdomain.blog domains include the param
- Test mapped.com domains include the param
- Test wpcomstaging urls do not include the param
- Test domain redirects do not include the param
- Test Atomic sites do not include the param

For testing in calypso using chrome with third party cookies disabled you'll need to use calypso's oauth mode, e.g. `http://calypso.localhost:3000/domains/manage/testing23465324891.home.blog?flags=oauth,logmein`.

Unit tests:
```
yarn test-client client/lib/logmein
```

Related to #52741 / #52740 / #52763

aneeshd16 commented:
Adding a link on the domain management page would also work!
![image](https://user-images.githubusercontent.com/5436027/118505955-84e51280-b74a-11eb-8a6c-da4797a165ee.png)

https://github.com/Automattic/wp-calypso/blob/e9998f7ffff930b477fb32dddf4fc87530d80be6/client/my-sites/domains/domain-management/list/index.jsx#L401

aneeshd16 commented on client/components/web-preview/toolbar.jsx:
```diff
@@ -181,7 +182,7 @@ class PreviewToolbar extends Component {
 							primary
 							className="web-preview__external"
-							href={ externalUrl || previewUrl }
+							href={ logmeinUrl( externalUrl ) || previewUrl }
```
We should probably do this change for only WPCOM sites since Jetpack sites won't recognize this param. What do you think?

lsl replied:
Good point, do you think we can leverage the mysites data for this?

aneeshd16 replied:
We can probably just use this selector:
https://github.com/Automattic/wp-calypso/blob/trunk/client/state/selectors/is-site-wpcom.js

niranjan-uma-shankar commented on packages/logmein/src/index.ts:
```diff
@@ -0,0 +1,26 @@
+		return fullUrl;
+	}
+
+	url.searchParams.set( 'logmein', '1' );
```
Instead of appending logmein param, would it make sense to directly take a logged in user to the the r-login.wordpress.com endpoint? If the user is in Calypso and logged in, we don't need to take them to wordpress.com/log-in - we can take them to `https://r-login.wordpress.com/remote-login.php?action=link&back=https://{MAPPED_DOMAIN.tld}` and save the extra log in screen step:

<img width="1032" alt="Screenshot_2021-05-21_at_2_34_34_AM" src="https://user-images.githubusercontent.com/1269602/119057252-c47d5a00-b9e9-11eb-8245-ed04ea5f6762.png">

Appending logmein=1 is okay if we find the user is logged out for some reason. So, I'm thinking:

```
if ( isLoggedIn ) {
  navigate to https://r-login.wordpress.com/remote-login.php?action=link&back=https://{MAPPED_DOMAIN.tld}
} else {
  navigate to https://mapped-domain.tld/?logmein=1
}
```

A scenario in which this may not work is if the user has Calypso open in a tab, logs out of WPCOM in a different tab, and clicks on "Visit site" button on the open Calypso tab. Then they'll still be logged in to the mapped domain even though they're not signed in to WPCOM. Probably there are some checks we can add to handle this.

lsl replied:
This should be fine, I'll see what I can do.

lsl replied:
I tried this here: https://github.com/Automattic/wp-calypso/pull/52909/files#diff-d025c6a21229fe69836222916356ee8a2c395861252a82ff88e97b9facc70a4bR31

But it didn't seem to work for some reason (didn't dig too deep) end up getting js output as plaintext directly from `remote-login.php

niranjan-uma-shankar reviewed (commented):
Should we also add these logmein links to the editor?

<img width="1527" alt="Screenshot 2021-05-21 at 1 54 46 PM" src="https://user-images.githubusercontent.com/1269602/119119544-e954ea00-ba48-11eb-86fc-9ecfeb3993cb.png">

<img width="1511" alt="Screenshot 2021-05-21 at 1 55 00 PM" src="https://user-images.githubusercontent.com/1269602/119119671-0984a900-ba49-11eb-94eb-6888dea7335a.png">

A couple of other places:

In themes page after activating a theme:

<img width="1410" alt="Screenshot 2021-05-21 at 1 59 13 PM" src="https://user-images.githubusercontent.com/1269602/119120231-acd5be00-ba49-11eb-8848-9ee3170a0544.png">

In the Reader where the title links to the mapped domain site:

<img width="1343" alt="Screenshot 2021-05-21 at 2 00 40 PM" src="https://user-images.githubusercontent.com/1269602/119120364-d262c780-ba49-11eb-970b-50aada699b70.png">

lsl committed:
Add logmeinUrl as lib

lsl committed:
Add getSitesMappedDomains selector,

and use it with logmeinUrl to apply logmein params to domain links.

lsl commented:
Reworked this as a client lib for simplicity and updated code to match up with what @JoshuaGoode is dropping in the demo. Depends on the latest update in D61461-code.

Edit: Just noticed I nuked the tests in the process, will add them back tomorrow and amend to include the new selector. (Maybe, if we're still going this way)

lsl committed:
Add tests back

lsl committed:
Test fixes, mv

niranjan-uma-shankar commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,44 @@
+
+	// return (
+	// 	'https://r-login.wordpress.com/remote-login.php?action=link&back=' +
+	// 	encodeURIComponent( fullUrl )
```
This should work, since I have a similar implementation in https://github.com/Automattic/wp-calypso/pull/53125/ and it works well. One thing to note is that the back URL should be `https` instead of `http`. When I tested this PR by clicking on the mapped domain link on the domain management page, I was taken to the following URL which [outputted JS on the screen](https://d.pr/i/KbqF2a):

```
https://r-login.wordpress.com/remote-login.php?action=link&back=http%3A%2F%2Fniranjan792177516.blog
```



Note the `http` in the back URL. When I changed that `https`, then the r-login redirect worked well. 

Anyhow, with the change in D61461-code to skip the Continue step, is this change still needed?

niranjan-uma-shankar commented on client/my-sites/domains/domain-management/list/index.jsx:
```diff
@@ -398,7 +400,7 @@ export class List extends React.Component {
 					<ExternalLink
 						className="list__header-primary-domain-url"
-						href={ selectedSite.URL }
+						href={ logmeinUrl( selectedSite.URL, siteUrls ) }
```
Would you like to include in this PR the logmeinUrl()'ification of other instances in Calypso that link to the mapped domain URL? We might have to do this in the editor, reader and so on. Or do you think we should take that as a separate PR?

lsl committed:
Rename mapped domains selector to be more specific, shore up logmein https, direct flag value

lsl committed:
Default logged in and direct, test fixes

lsl commented on client/state/selectors/get-sites-logmein-allowed-urls.js:
```diff
@@ -0,0 +1,29 @@
+				: false
+		)
+		.filter( Boolean );
+	return result;
```
I'm not sure if this is the correct list of restrictions.. do we have an "is simple site" method by chance?

niranjan-uma-shankar replied:
I don't know the right answer to this, but here's what I found. 
We have [this selector](https://github.com/Automattic/wp-calypso/blob/b703449c2b079916da70356acd66ca823760cebf/client/state/selectors/is-site-wpcom.js) that returns true if a site is a WPCOM site(atomic or simple) as described the JS Doc:

> Whether the site is a WPCOM site ( Atomic or Simple )

Using the logic in that selector, the following should return simple sites:

```
! isJetpackSite( state, siteId ) || ! isAtomicSite( state, siteId )
```

Would this be the right check for WPCOM simple sites?

lsl commented:
Moving focus to replacement: #53443
